### PR TITLE
1.17 support

### PR DESCRIPTION
**Pull request by nickelpro**
This PR is for coordinating the integration of 1.17. Please do not report "test fail" or "doesn't work" on this PR, we're aware and it just hurts the signal to noise ratio. This PR is not ready to merge, tests are broken on every version except 1.17. Current focus of development is integrating the old transaction system with the new `set_slot` based system. If you're wondering about what needs to be done more specifically, please drop by the Discord.

Co-authored-by: @u9g 
Co-authored-by: @willocn
Co-authored-by: @imharvol 
Co-authored-by: @rom1504 
Co-authored-by: @extremeheat 
Co-authored-by: @Archengius

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1979: 1.17 support](https://issuehunt.io/repos/1284185/issues/1979)
---
</details>

### CODE REVIEW

File: lib/plugins/digging.js [hunk 1]
```diff
@@ -176,15 +176,39 @@ function inject (bot) {
     return block && block.diggable && block.position.offset(0.5, 0.5, 0.5).distanceTo(bot.entity.position.offset(0, 1.65, 0)) <= 5.1
   }
 
+  function getItemEnchantments (item) {
```
Comment (u9g) [inline lib/plugins/digging.js:179]: there's a getter for this already on prismarine-item objects already .enchants

File: lib/plugins/entities.js [hunk 2]
```diff
@@ -31,6 +31,11 @@ function inject (bot, { version }) {
   const entitiesArray = require('minecraft-data')(version).entitiesArray
   const Item = require('prismarine-item')(version)
   const ChatMessage = require('prismarine-chat')(version)
+  // ONLY 1.17 has this destroy_entity packet which is the same thing as entity_destroy packet except the entity is singular
+  // 1.17.1 reverted this change so this is just a simpler fix
+  bot._client.on('destroy_entity', (packet) => {
+    bot._client.emit('entity_destroy', { entityIds: [packet.entityId] })
```
Comment (u9g) [inline lib/plugins/entities.js:37]: needs to be refactored

**Comment by nickelpro** [comment]
1.17 test failures are due to race conditions in trade and useChests tests, investigating

**Comment by nickelpro** [comment]
To be clear, useChests def still has a race condition in it

File: test/externalTests/useChests.js [hunk 3]
```diff
@@ -91,10 +101,10 @@ module.exports = () => async (bot) => {
       assert(blockAssert && block2Assert, new Error('The block instance emitted by chestLidMove is not part of the chest oppened'))
       assert.strictEqual(isOpen, 1, new Error('isOpen should be 1 when opened by one only player'))
 
-      setTimeout(() => {
-        bot.removeListener('chestLidMove', handler)
-        chest.close()
-      }, 500)
+      await bot.test.wait(500)
```
Comment (nickelpro) [inline test/externalTests/useChests.js:104]: Unnecessary on 1.17 and wrong on older versions

File: test/externalTests/team.js [hunk 4]
```diff
@@ -12,7 +13,7 @@ module.exports = () => async (bot) => {
     bot.test.sayEverywhere('/scoreboard teams join test')
   }
 
-  await once(bot, 'teamMemberAdded')
+  await p
```
Comment (nickelpro) [inline test/externalTests/team.js:16]: p is pointless here, just await 'teamMemberAdded'

File: test/externalTests/plugins/testCommon.js [hunk 5]
```diff
@@ -33,9 +33,14 @@ function inject (bot) {
...
-    const p = bot.awaitMessage('Block placed', /Changed the block/)
+    const { x: _x, y: _y, z: _z } = relative ? bot.entity.position.floored().offset(x, y, z) : { x, y, z }
+    const block = bot.blockAt(new Vec3(_x, _y, _z))
+    if (block.name === blockName) {
+      return
+    }
+    const p = once(bot.world, `blockUpdate:(${_x}, ${_y}, ${_z})`)
     const prefix = relative ? '~' : ''
     bot.chat(`/setblock ${prefix}${x} ${prefix}${y} ${prefix}${z} ${blockName}`)
     await p
```
Comment (nickelpro) [inline test/externalTests/plugins/testCommon.js:46]: another pointless `p`, `.chat()` doesn't await anything, so await the block update after chat

File: test/externalTests/crafting.js [hunk 6]
```diff
@@ -49,4 +50,9 @@ module.exports = () => async (bot) => {
   await bot.test.setInventorySlot(36, new Item(populateBlockInventory.id, 1, 0))
   await bot.test.becomeSurvival()
   await craft(1, craftItem)
+  await bot.test.setBlock({ x: 1, y: 0, z: 0, relative: true, blockName: 'crafting_table' })
+  bot.chat('/give @p stick 7')
+  await once(bot.inventory, 'updateSlot')
+  const craftingTable = bot.findBlock({ matching: blocksByName.crafting_table.id })
+  await bot.craft(bot.recipesFor(itemsByName.ladder.id, null, null, true)[0], 1, craftingTable)
```
Comment (nickelpro) [inline test/externalTests/crafting.js:57]: We need to test the results of these crafting interactions

File: lib/plugins/inventory.js [hunk 7]
```diff
@@ -88,11 +94,7 @@ function inject (bot, { version, hideErrors }) {
       const item = window.findItemRange(start, end, window.selectedItem.type, window.selectedItem.metadata, true, window.selectedItem.nbt)
 
       if (item && item.stackSize !== item.count) { // something to join with
-        if (noWaiting) {
-          await clickWindow(item.slot, 0, 0)
-        } else {
-          await Promise.all([clickWindow(item.slot, 0, 0), once(window, `updateSlot:${item.slot}`)])
-        }
+        await clickWindow(item.slot, 0, 0)
```
Comment (nickelpro) [inline lib/plugins/inventory.js:97]: `noWaiting` never did what it advertised it did, and now that we don't use it at all we should remove it

File: test/externalTests/useChests.js (hunk 3 above)
Comment (imharvol) [reply test/externalTests/useChests.js:104]: The idea is to keep the chest open for a litle bit to make sure that no more 'chestLidMove' are emitted until it's closed

Comment (nickelpro) [reply test/externalTests/useChests.js:104]: Ahhh, I forgot that it was a time dependent thing. Thanks for the heads up

**Comment by rom1504** [comment]
I see 1.17 tests are passing, nice!

File: test/externalTests/breath.js [hunk 8]
```diff
@@ -2,5 +2,6 @@ const assert = require('assert')
 
 module.exports = () => async (bot) => {
   await new Promise((resolve, reject) => setTimeout(resolve, 5000))
-  assert.strictEqual(bot.oxygenLevel, 20, 'Wrong oxygen level')
+  // TODO: ** Fix in 1.17
+  if (bot.oxygenLevel) assert.strictEqual(bot.oxygenLevel, 20, 'Wrong oxygen level')
```
Comment (extremeheat) [inline test/externalTests/breath.js:6]: I think this should work now, can re-enable it

File: lib/plugins/inventory.js [hunk 9]
```diff
@@ -333,10 +333,63 @@ function inject (bot, { version, hideErrors }) {
...
+      }
+      return true
+    }
+    return false
+  }
+
+  async function waitForWindowUpdate (window, slot) {
+    if (window.type === 'minecraft:inventory') {
+      if (slot >= 1 && slot <= 4) {
+        await once(bot.inventory, 'updateSlot:0')
```
Comment (nickelpro) [inline lib/plugins/inventory.js:373]: This is insufficient for several reasons, need code close to `expectTradeUpdate`

File: lib/plugins/inventory.js [hunk 10]
```diff
@@ -401,27 +454,56 @@ function inject (bot, { version, hideErrors }) {
...
-
-    const [success] = await response
-    if (!success) {
-      throw new Error(`Server rejected transaction for clicking on slot ${slot}, on window ${JSON.stringify(window.slots, null, 2)}.`)
+      const [success] = await response
+      if (!success) {
+        throw new Error(`Server rejected transaction for clicking on slot ${slot}, on window ${JSON.stringify(window.slots, null, 2)}.`)
+      }
+    } else {
+      await waitForWindowUpdate(window, slot)
```
Comment (nickelpro) [inline lib/plugins/inventory.js:506]: Can any other versions benefit from `waitForWindowUpdate`? Needs investigation

File: lib/plugins/place_block.js [hunk 11]
```diff
@@ -4,13 +4,14 @@ module.exports = inject
...
+    await bot._genericPlace(referenceBlock, faceVector, options)
 
-    const dest = pos.plus(faceVector)
-    const eventName = `blockUpdate:${dest}`
-
-    const [oldBlock, newBlock] = await onceWithCleanup(bot, eventName, { timeout: 5000 })
+    let newBlock = bot.blockAt(dest)
+    if (oldBlock.type === newBlock.type) {
+      [oldBlock, newBlock] = await onceWithCleanup(bot, `blockUpdate:${dest}`, { timeout: 5000 })
+    }
```
Comment (nickelpro) [inline lib/plugins/place_block.js:14]: This approach is overly complicated, switch to @u9g's suggestion of just awaiting both `_genericPlace` and `blockUpdate` together if possible

File: lib/plugins/villager.js [hunk 12]
```diff
@@ -127,76 +128,69 @@ function inject (bot, { version }) {
...
-      if (slot2) {
-        assert.strictEqual(slot2.type, Trade.inputItem2.type)
-        const updatedCount2 = slot2.count - Trade.inputItem2.count
-        const updatedSlot2 = updatedCount2 <= 0
-          ? null
-          : { ...slot2, count: updatedCount2 }
-        villager.updateSlot(1, updatedSlot2)
-      }
+      // ToDo: See if this does anything kappa
+      // villager.updateSlot(2, Object.assign({}, Trade.outputItem))
```
Comment (nickelpro) [inline lib/plugins/villager.js:181]: Resolve this ToDo before merge

File: package.json [hunk 13]
```diff
@@ -26,12 +26,12 @@
...
     "prismarine-block": "^1.6.0",
     "prismarine-chat": "^1.1.0",
-    "prismarine-chunk": "^1.20.3",
+    "prismarine-chunk": "^1.24.0",
     "prismarine-entity": "^1.0.0",
     "prismarine-item": "^1.6.0",
     "prismarine-physics": "^1.0.4",
     "prismarine-recipe": "^1.1.0",
-    "prismarine-windows": "^2.2.0",
+    "prismarine-windows": "github:nickelpro/prismarine-windows",
```
Comment (nickelpro) [inline package.json:34]: Merge pr-window and switch this back to npm before merge

**Comment by nickelpro** [comment]
All force pushes are just rebasing onto master to get the branch fresh

**Comment by imharvol** [comment]
I'm not so sure that there's a race condition on useChests. Could the trade test failing be influencing the useChests test?

I've run the useChests multiple times (`npm run mocha_test -- -g useChests`) and it's not failing, but when I also include the trade tests (`npm run mocha_test -- -g "(trade|useChests)"`), both of them fail

And if you see the order of the tests, the useChests test is the only one after trade tests
image

I tried running trade as the first test and the result is that many other tests also fail
[log-tests.txt](https://github.com/PrismarineJS/mineflayer/files/6999080/log-tests.txt)

**Comment by nickelpro** [comment]
It does, but  don't take my word for it. I'm going to fix the race in trade tonight and the useChests will be the only failing test.

File: lib/plugins/villager.js [hunk 14]
```diff
@@ -178,29 +178,31 @@ function inject (bot, { version }) {
     for (let i = 0; i < count; i++) {
       await putRequirements(villager, Trade, count)
       // ToDo: See if this does anything kappa
-      villager.updateSlot(2, Object.assign({}, Trade.outputItem))
       Trade.nbTradeUses++
       if (Trade.maximumNbTradeUses - Trade.nbTradeUses === 0) {
         Trade.tradeDisabled = true
       }
+      if (!bot.supportFeature('setSlotAsTransaction')) {
```
Comment (nickelpro) [inline lib/plugins/villager.js:185]: I suspect this isn't necessary but it's hard to say with tests still broken

File: lib/plugins/generic_place.js [hunk 15]
```diff
@@ -37,7 +38,7 @@ function inject (bot, { version }) {
       await bot.lookAt(referenceBlock.position.offset(dx, dy, dz), options.forceLook)
     }
     // TODO: tell the server that we are sneaking while doing this
-    const pos = referenceBlock.position
+    const pos = referenceBlock.position.plus(faceVector)
```
Comment (imharvol) [inline lib/plugins/generic_place.js:41]: This is probably what broke 1.8 sign tests. You should not send `.plus(faceVector)` on the packet, and the methods that use _genericPlace already do this, so no need to return it either
https://github.com/PrismarineJS/mineflayer/blob/ec683de1f714db593c4843bddf8e45595eb637ea/lib/plugins/place_entity.js#L48
https://github.com/PrismarineJS/mineflayer/blob/ec683de1f714db593c4843bddf8e45595eb637ea/lib/plugins/place_block.js#L10

Comment (nickelpro) [reply lib/plugins/generic_place.js:41]: The reference block is not the position you should send in the packet, the packet position needs to the position you're _placing the block in_. If you use the reference block position here you end up trying to put the block _into_ an existing block. The furnace test bug is linked to this problem.

Those `dest` calculations were added by me for the purpose of fixing race conditions with relying on _generic_place to do the calculation, but the API pre-dates this PR and I didn't want to break it.

Comment (nickelpro) [reply lib/plugins/generic_place.js:41]: You're correct about this btw, thanks

**Comment by rom1504** [comment]
anyone feels like checking the main example ?
to make sure nothing big was missed

**Comment by rom1504** [comment]
Alright it's passing.

Left to do:
* Check if any doc needs to be updated
* check if main examples work ok
* Any clean up to do ?
* Merge and enjoy

**Comment by rom1504** [comment]
Also:
* Does anyone want a test for 1.17 in addition than 1.17.1 ?
* can everyone have a look at the changes and try to catch any possible issue ?

**Comment by rom1504** [comment]
Ok I looked at the code, I don't see any obvious issue, LGTM
Looks like no doc needs updating.

So the only thing left (besides possible cleaning @nickelpro may want to do but imo that's not blocking) is checking some examples.

**Comment by rom1504** [comment]
* [ ] inventory.js
* [ ] chest.js
* [ ] digger.js
* [ ] jumper.js

**Comment by nevercast** [comment]
Testing jumper.js now

**Comment by nickelpro** [comment]
My cleanup is not blocking, it can get in at any time

**Comment by u9g** [comment]
> jumper.js tested.
> 
> * ✔️ Jumping and moving is fine, mounting works for Boat and Minecart. Minecart can be propelled forward and backward.
> * ❌ Boat is immobile.
> * ✔️ Attacking works
> * ⚠️ Can't mount horse, idk if that ever worked.
> 
> digger.js works and all functionality operates. Occasional promise rejections cause the process to crash.
> `(node:11276) UnhandledPromiseRejectionWarning: Error: Event blockUpdate:(93, 8, -121) did not fire within timeout of 5000ms` occurred while spamming `dig`

while spamming dig, did you await the digs?

**Comment by nevercast** [comment]
> while spamming dig, did you await the digs?

https://github.com/PrismarineJS/mineflayer/blob/master/examples/digger.js#L67 

Nope I did not.

File: test/externalTests/crafting.js (hunk 6 above)
Comment (u9g) [reply test/externalTests/crafting.js:57]: done in my pr that will be merged after this

**Pull request merged**